### DISCUSSION TITLE
What is the `openjdk` formula, actually?

### DISCUSSION THREAD

**Discussion by bartekpacia** [Homebrew/discussions#5491] [Everyday usage] [answered]
### Output of `brew config`

```shell
(doesn't matter)
```

### Output of `brew doctor`

```shell
(doesn't matter)
```

### Description of issue

Hi, first of all - thank you to everyone maintaining and contributing to Homebrew. I love it.

Now, onto my question: I'm not sure _what_ is the [`openjdk`](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/o/openjdk.rb) formula. Is this a distribution of Java by Homebrew? Is it just a upstream build?

I know that there are distributins of JDKs (e.g. Temurin by Eclipse, Amazon Coretto, even Microsoft have one). E.g. the official [`setup-java`](https://github.com/actions/setup-java) GitHub action allows for specifying many distributions, but `openjdk` is not one of them:

<img width="758" alt="Screenshot 2024-07-14 at 12 48 58" src="https://github.com/user-attachments/assets/33bec99f-f2a6-4e52-9af4-a58fd843363b">

Another, but related question: why is [`openjdk`](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/o/openjdk.rb) a formula, but e.g. [`temurin`](https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/t/temurin.rb) is a cask?

I'm sorry if my question doesn't necessarily makes sense; in that case I'd be grateful for pointing out flaws in my understanding. Thank you.

**Comment by SMillerDev** [comment, marked as answer]
> Now, onto my question: I'm not sure what is the [openjdk](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/o/openjdk.rb) formula. Is this a distribution of Java by Homebrew? Is it just a upstream build?

Every formula is a build of the software upstream provides, compiled on Homebrew CI. In the OpenJDK case: https://openjdk.org/ is upstream. And everything in the list you posted are builds of it by different vendors.

> Another, but related question: why is [openjdk](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/o/openjdk.rb) a formula, but e.g. [temurin](https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/t/temurin.rb) is a cask?

Because a formula is an upstream project that can (and will) be compiled by brew. And a cask is a Homebrew distribution of precompiled binaries, like the temurin project or Firefox provides.

**Reply by bartekpacia** [reply]
I see, this confirms my suspicions. Thank you.
